### PR TITLE
ci: factor out go mod cache permissions fix

### PR DESCRIPTION
**Pull request by samsalisbury**
Eliminates some repetition.

UPDATE: This needs some work, as it's causing issues on the non-machine executors due to lack of circleci user. I'll get to this tomorrow.

### CODE REVIEW

File: .circleci/config/@config.yml [hunk 1]
```diff
@@ -24,6 +24,12 @@ commands:
             - ui/node_modules
   restore_go_cache:
     steps:
+      - run:
+          name: Allow circleci user to restore Go modules cache
+          command: |
+            set -eux -o pipefail
+            sudo mkdir /go
```
Comment (nathanhruby) [inline .circleci/config/@config.yml]:
```suggestion
            sudo mkdir -p /go
```

Comment (nathanhruby) [reply .circleci/config/@config.yml]: also the go executor doesn't have sudo (just the go-machine one) so might need to switch the go-mod-download to a machine executor

Comment (samsalisbury) [reply .circleci/config/@config.yml]: Thanks for the suggestion @nathanhruby! Good catch on the sudo as well, I'll think about how to approach this as would rather avoid using the machine executors except where really necessary.

Comment (nathanhruby) [reply .circleci/config/@config.yml]: `which sudo ; if [ $? -eq 0 ] ; then SUDO=sudo ; else SUDO="" ; fi` should work

Comment (samsalisbury) [reply .circleci/config/@config.yml]: Went for a slight variation on that, as I think `which sudo` would have caused failure on docker, as we use `set -e` amongst other options. See https://github.com/hashicorp/vault/pull/6894/files#diff-f3896cfad122301481e66fb84d437980R31

Comment (samsalisbury) [reply .circleci/config/@config.yml]: Nope, still issues and I need to finish my day. I'll fix this tomorrow.

```
++ command -v sudo
+ SUDO=
+ mkdir -p /go
+ chown -R circleci:circleci /go
chown: invalid user: 'circleci:circleci'
Exited with code 1
```

**Pull request closed without merging**